---
title: "Bioinformatics in R Final Project: Does CBD impact HIV+ Patients?"
author: "Matthew Kirschenbaum"
date: "May 7, 2024"
output: 
  html_document: 
    highlight: espresso
    theme: cerulean
---

```{r setup}
    ### Edit this line so your notebook renders properly
    knitr::opts_knit$set(root.dir = normalizePath("/Users/matthewkirschenbaum/Desktop/Bioinformatics in R")) 
```


Load the Libraries
```{r message=FALSE, warning=FALSE, paged.print=FALSE, results="hide"}
library(dplyr)
library(readr)
library(DESeq2)
library(ggplot2)
library(EnhancedVolcano)
```

Upload Counts table and Metadata
```{r}
counts <- read.table('/Users/matthewkirschenbaum/Desktop/Bioinformatics in R/counts_table.txt', header = TRUE, sep = '\t')
metadata <- read.csv('finalproject_metadata.csv', row.names = 1, header = TRUE)
```

Subset the Data and Match Column Names
```{r}
# Set the first column ('Geneid') as row names
rownames(counts) <- counts[, 1]

# Select relevant columns (7 to 12) and rename them according to our samples
counts_subset <- counts[, 7:12]
colnames(counts_subset) <- c("SRR24035598", "SRR24035597", "SRR24035596", "SRR24035595", "SRR24035594", "SRR24035593")
```

Run DESeq on subsetted data with respect to metadata and CBD treatment
```{r}
dds <- DESeqDataSetFromMatrix(countData = counts_subset, 
                              colData = metadata,
                              design = ~ Treatment)
```

Subset the DESeq Results
```{r}
## Subset your ddseq object
subset <-  dds[(rowSums(counts(dds)) >= 10),]

### fix the base to a  treatment
relevel(dds$Treatment, ref='Baseline')

deseq_ob <- DESeq(subset)
res <- results(deseq_ob, alpha = 0.05)
```

Rename Differentially Expressed Ensembl Genes using org.Hs.eg database
```{r}
library(org.Hs.eg.db)

sigs.df <-  as.data.frame(res)
enst <- rownames(sigs.df)
enst_adjusted <- gsub("\\..*$", "", enst) #remove the decimal and numbers after the decimal in the Ensembl geneID row names for compatibility with org.Hs.eg.db
sigs.df$symbol <- mapIds(org.Hs.eg.db, keys= enst_adjusted, keytype = 'ENSEMBL', colum = "SYMBOL")
```

Visualize with Standard Volcano Plot
```{r}
EnhancedVolcano(sigs.df, x='log2FoldChange', y = 'padj', lab = sigs.df$symbol)
```

Make Volcano Plot Vertical to emphasize p-value difference
```{R}

EnhancedVolcano(res,
  lab = sigs.df$symbol,
  x = 'log2FoldChange',
  y = 'pvalue',
  xlab = bquote(~Log[2]~ 'fold change'),
  pCutoff = 10e-14,
  FCcutoff = 1.0,
  pointSize = 3.0,
  labSize = 6.0,
  labCol = 'black',
  labFace = 'bold',
  boxedLabels = TRUE,
  parseLabels = TRUE,
  col = c('black', 'pink', 'purple', 'red3'),
  colAlpha = 4/5,
  legendPosition = 'bottom',
  legendLabSize = 14,
  legendIconSize = 4.0,
  drawConnectors = TRUE,
  widthConnectors = 1.0,
  colConnectors = 'black') + coord_flip()  # Add 'coord_flip()' to make the plot vertical

```

Heatmap of Differentially-Expressed Genes by Sample
```{R}
library(ComplexHeatmap)
diff.df <- as.data.frame(sigs.df)
diff.df <- diff.df %>%
  filter(padj < 0.05)

mat <- counts(deseq_ob, normalized = T)[rownames(diff.df),]
mat.z <- t(apply(mat,1, scale))
colnames(mat.z) <- colnames(mat)
Heatmap(mat.z, cluster_rows= T, cluster_columns= T, name = "Z-score", row_labels = diff.df[rownames(mat.z),]$symbol)
```

Gene Ontology Analysis
```{R}
library(clusterProfiler)

diff.df <- as.data.frame(sigs.df)
diff.df <- diff.df %>%
  filter(padj < 0.05)
diff_df_adjusted <- diff.df

# Extract rownames
rownames_diff_df <- rownames(diff.df)

# Remove decimal or number after decimal for use with org.Hs.eg.db
rownames_adjusted <- gsub("\\..*$", "", rownames_diff_df)

# Assign adjusted rownames to the new dataframe
rownames(diff_df_adjusted) <- rownames_adjusted

genes_to_test <- rownames(diff_df_adjusted)
GO_results <- enrichGO(gene = genes_to_test, OrgDb = "org.Hs.eg.db", keyType = 'ENSEMBL', ont = "BP")
fit <- plot(barplot(GO_results, showCategory=10))
```

GOSt Plot
```{R}
library(gprofiler2) 
go_upreg <- gost(genes_to_test, organism = 'hsapiens', significant = T)
gostplot(go_upreg, interactive = T)
```


GSEA Plot: Works in RStudio Markdown, but not when rendering/knitting. Included code for reference.
```{R}
res <- res[order(-res$stat),]
gene_list <- res$stat
names(gene_list) <- rownames(res)
gse <- gseGO(gene_list, ont = "BP", keyType = "ENSEMBL", OrgDb ="org.Hs.eg.db", eps = 1e-300 )
gseaplot(gse, geneSetID = "GO:0046324" )

```
